### PR TITLE
REST Client Reactive - logging requests and responses

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
Example log lines:

```
1-11-08 20:36:26,828 DEBUG [org.jbo.res.rea.cli.imp.DefaultClientLogger] (vert.x-eventloop-thread-47) Request: POST http://localhost:8081/hello Headers[Accept=text/plain;charset=UTF-8 Content-Type=application/json content-length=5], Body:
Hello

2021-11-08 20:36:26,868 DEBUG [org.jbo.res.rea.cli.imp.DefaultClientLogger] (vert.x-eventloop-thread-47) Redirect: POST http://localhost:8081/hello, Status[303 See Other], Headers[Content-Type=text/plain;charset=UTF-8 Location=http://example.com content-length=0], Body:

2021-11-08 20:36:27,120 DEBUG [org.jbo.res.rea.cli.imp.DefaultClientLogger] (vert.x-eventloop-thread-47) Response: GET http://example.com/, Status[200 OK], Headers[Age=366555 Cache-Control=max-age=604800 Content-Type=text/html; charset=UTF-8 Date=Mon, 08 Nov 2021 19:36:27 GMT Etag="3147526947+ident" Expires=Mon, 15 Nov 2021 19:36:27 GMT Last-Modified=Thu, 17 Oct 2019 07:18:26 GMT Server=ECS (nyb/1D13) Vary=Accept-Encoding X-Cache=HIT content-length=1256], Body:
<!doctype html>...
```

refs #21286

### CODE REVIEW

File: extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java [hunk 1]
```diff
@@ -45,4 +45,21 @@
...
+     * <ul>
+     * <li>{@code request-response} - enables logging request and responses, including redirect responses</li>
+     * <li>{@code all} - enables logging requests and responses and lower-level logging</li>
+     * <li>{@code none} - no additional logging</li>
+     * </ul>
+     *
+     * This property is applicable to reactive REST clients only.
+     */
+    @ConfigItem
+    public Optional<String> logLevel;
```
Comment (geoand) [inline extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java:63]: Shouldn't this use the `LogLevel` enum?

Comment (michalszynkiewicz) [reply extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java:63]: the problem is the `RestClientsConfig` class is shared between rest-client and rest-client-reactive, so doesn't have access to RESTEasy Reactive stuff. It's the same for `multipartPostEncoderMode` above

Comment (geoand) [reply extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java:63]: Gotcha, thanks

**Review by geoand — APPROVED** [review]

**Comment by maxandersen** [comment]
looks good - I'm wondering if we really want to have the full body printed into the logs ? maybe just include the first 100 characters and if there is more put "..." ?

maybe have a way to set the limit to something higher if you really want your logs to be massive :)

File: extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java [hunk 2]
```diff
@@ -45,4 +45,21 @@
...
+     * The possible values are:
+     * <ul>
+     * <li>{@code request-response} - enables logging request and responses, including redirect responses</li>
+     * <li>{@code all} - enables logging requests and responses and lower-level logging</li>
+     * <li>{@code none} - no additional logging</li>
+     * </ul>
+     *
+     * This property is applicable to reactive REST clients only.
+     */
+    @ConfigItem
```
Comment (maxandersen) [inline extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java:62]: is "logLevel" the right term ? we use "log.level" for debug/info/error/tracing else where which is not the same ....this here is more about "detailLevel" ?

Comment (michalszynkiewicz) [reply extensions/resteasy-classic/rest-client/config/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java:62]: maybe `quarkus.rest-client.logging.scope` ?

Change from `-` to `.` would also open a possibility to have properties further customizing logging

**Comment by michalszynkiewicz** [comment]
> looks good - I'm wondering if we really want to have the full body printed into the logs ? maybe just include the first 100 characters and if there is more put "..." ?
> 
> maybe have a way to set the limit to something higher if you really want your logs to be massive :)

That may be a good start of further improvements. 
I'll change the PR to use a different property but would like to add things like more configuration options in a next PR.

I'm also thinking of letting users use their own [`ClientLogger`](https://github.com/quarkusio/quarkus/pull/21286/files#diff-fc30c990b48f0a314bf4dbd0a9d94bdab74f505dd915a21b1e5c8ab464b28d9bR7), this way they can e.g. skip logging sensitive data (auth headers, etc)

**Comment by michalszynkiewicz** [comment]
> looks good - I'm wondering if we really want to have the full body printed into the logs ? maybe just include the first 100 characters and if there is more put "..." ?

I ended up changing `log-level` to `logging.scope` and adding `logging.body-limit` :)

**Comment by rsvoboda** [comment]
@michalszynkiewicz this is nice feature, but nobody knows about it :) Can you please enhance the docs to point it out?

Also didn't find the properties on https://quarkus.io/guides/all-config

**Comment by michalszynkiewicz** [comment]
@rsvoboda I created a PR to add some info to the guide: https://github.com/quarkusio/quarkus/pull/22445

**Comment by geoand** [comment]
@deejonz see [this](https://github.com/quarkiverse/quarkus-langchain4j/blob/0.18.0.CR1/model-providers/openai/openai-common/runtime/src/main/java/io/quarkiverse/langchain4j/openai/QuarkusOpenAiClient.java#L95) example.

**Comment by geoand** [comment]
That builder doesn't give you that ability

**Comment by JPSantistebanQ** [comment]
The `DefaultClientLogger` print `DEBUG` logs.
How can I change that, for example to `INFO`?

**Comment by geoand** [comment]
```
quarkus.log.category."org.jboss.resteasy.reactive.client.logging".level=DEBUG
```

**Comment by geoand** [comment]
You would need your own `ClientLogger`

**Pull request merged**